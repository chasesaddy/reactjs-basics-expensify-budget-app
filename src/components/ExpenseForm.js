import React from 'react';

import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

class ExpenseForm extends React.Component {
  state = {
    description: '',
    note: '',
    amount: '',
    createdAt: moment(),
    calFocused: false,
    errorMsg: ''
  };

  onDescriptionChange = ( e ) => {
    const description = e.target.value;
    this.setState( () => ( { description } ) );
  };

  onNoteChange = ( e ) => {
    const note = e.target.value;
    this.setState( () => ( { note } ) );
  };

  onAmountChange = ( e ) => {
    const amount = e.target.value;
    if ( !amount || amount.match(/^\d{1,}(\.\d{0,2})?$/) ) {
      this.setState( () => ( { amount } ) );
    }
  };

  onDateChange = ( createdAt ) => {
    if ( createdAt ) {
      this.setState( () => ( { createdAt } ) );
    }
  };

  onFocusChange  = ( { focused } ) => {
    this.setState( () => ( { calFocused: focused } ) );
  };

  onSubmit = ( e ) => {
    e.preventDefault();

    if ( !this.state.description || !this.state.amount ) {
      this.setState( () => ( { error: 'Need description and amount, please.' } ) );
    } else {
      this.setState( () => ( { error: '' } ) );
      console.log( 'Submitted' );
    }
  }

  render() {
    return (
      <div>
        
          { this.state.error && 
          <div className="errorMsg">{ this.state.error }</div>
          }        

        <form onSubmit={ this.onSubmit }>

          <input 
            type="text"
            placeholder="Description"
            autoFocus
            value={ this.state.description }
            onChange={ this.onDescriptionChange }
          />

          <input 
            type="text"
            placeholder="Amount"
            value={ this.state.amount }
            onChange={ this.onAmountChange }
          />

          <SingleDatePicker 
            date={ this.state.createdAt }
            onDateChange={ this.onDateChange }
            focused={ this.state.calFocused }
            onFocusChange={ this.onFocusChange }
            numberOfMonths={ 1 }
            isOutsideRange={ () => { false } }
          />

          <textarea
            placeholder="Add a note for your expense (optional)"
            value={ this.state.note }
            onChange={ this.onNoteChange }
          >
          </textarea>

          <button>
            Add Expense
          </button>

        </form>

      </div>
    )
  };
}

export default ExpenseForm;
