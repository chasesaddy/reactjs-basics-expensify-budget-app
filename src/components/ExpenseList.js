import React from 'react';
import { connect } from 'react-redux';

const ExpenseList = ( props ) => (
  <div>
    <h1>
      Expense List
    </h1>

    <div>
      <ul>
        { props.filters.text }
        { props.expenses.map( ( solo ) => <li key={ solo.id }>{ solo.description }</li> ) }
      </ul>
    </div>
  </div>
);

const mapStateToProps = ( state ) => {
  return {
    expenses: state.expenses
    , filters: state.filters
  };
};

export default connect( mapStateToProps )( ExpenseList );
