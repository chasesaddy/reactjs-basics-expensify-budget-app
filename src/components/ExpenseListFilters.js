import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';

import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  state = {
    calFocused: null,
    focusedInput: null
  };

  onDatesChange = ( { startDate, endDate } ) => {
    this.props.setStartDate( startDate );
    this.props.setEndDate( endDate );

    this.setState( {
      focusedInput: startDate && !endDate ? 'startDate' : 'endDate'
    } );
  };

  onFocusChange = ( calFocused ) => {
    this.setState( () => ( { calFocused } ) );
  };

  onTextChange = ( e ) => {
    this.props.setTextFilter( e.target.value );
  }

  onSortChange = ( e ) => {
    if ( e.target.value === 'date' ) {
      this.props.sortByDate();
    } else if ( e.target.value === 'amount' ) {
      this.props.sortByAmount();
    }
  }


  render() {
    return (
      <div>
        <input 
          type="text" 
          value={ this.props.filters.text } 
          onChange={ this.onTextChange }
        />

        <select 
          value={ this.props.filters.sortBy }
          onChange={ this.onSortChange }
        >
          <option value="date">
            Date
          </option>
          <option value="amount">
            Amount
          </option>
        </select>
        <DateRangePicker 
          startDate={ this.props.filters.startDate }
          endDate={ this.props.filters.endDate }
          startDateId="start"
          endDateId="end"
          onDatesChange={ this.onDatesChange }
          focusedInput={ this.state.focusedInput }
          onFocusChange={ this.onFocusChange }
          showClearDates={ true }
          numberOfMonths={ 1 }
          isOutsideRange={ () => false }
        />
      </div>
    );
  }
};

const mapStateToProps = ( state ) => ( {
  filters: state.filters
} );

const MapDispatchToProps = ( dispatch ) => ( { 
  setStartDate: ( startDate ) => dispatch( setStartDate( startDate ) ),
  setEndDate: ( endDate ) => dispatch( setEndDate( endDate ) ),
  setTextFilter: ( text ) => dispatch( setTextFilter( text ) ),
  sortByDate: () => dispatch( sortByDate() ),
  sortByAmount: () => dispatch( sortByAmount() ),
} );

export default connect( mapStateToProps, MapDispatchToProps )( ExpenseListFilters );
