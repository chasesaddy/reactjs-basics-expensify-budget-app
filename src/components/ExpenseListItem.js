import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = ( { id, description, amount, createdAt } ) => (
  <div>
    <h3>
      <Link to={ `${ id }/edit` }>{ description }</Link>
    </h3>
    <p>{ amount } - { createdAt }</p>    
  </div>
);

export default ExpenseListItem;
