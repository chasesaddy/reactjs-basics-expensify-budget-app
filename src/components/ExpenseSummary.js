import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeraljs';

import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export class ExpenseSummary extends React.Component {
  plural() {
    return this.props.expenseCount > 1 ? 'expenses' : 'expense';
  }
  totalAmount() {
    return numeral( this.props.expensesTotal / 100 ).format( '$0,0.00' );
  }

  render() {
    return (
      <div>
        Viewing { this.props.expenseCount } { this.plural() } totalling { this.totalAmount() }
      </div>
    );
  }
}

const mapStateToProps = ( state ) => {
  const visibleExpenses = selectExpenses( state.expenses, state.filters );

  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal( visibleExpenses )
  };
};

export default connect( mapStateToProps )( ExpenseSummary );
