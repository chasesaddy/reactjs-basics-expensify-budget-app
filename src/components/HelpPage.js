import React from 'react';

const HelpPage = () => (
  <div>
    The help page.
  </div>
);

export default HelpPage;
