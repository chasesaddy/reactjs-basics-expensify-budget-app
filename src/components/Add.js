import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
  onSubmit = ( oneExpense ) => {    
    this.props.onSubmit( oneExpense );
    this.props.history.push( '/' );
  };

  render() {
    return (
      <div>
        <h1>
          Add Expense
        </h1>

        <ExpenseForm 
          onSubmit={ this.onSubmit }
        />
      </div>
    );
  };
}

const mapsDispatchToProps = ( dispatch ) => ( {
  onSubmit: ( expense ) => dispatch( addExpense ( expense ) )
} );

export default connect( undefined, mapsDispatchToProps )( AddExpensePage );
