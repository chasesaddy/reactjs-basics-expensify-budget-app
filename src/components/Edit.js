import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

const EditExpensePage = ( props ) => (
  <div>
    <ExpenseForm 
      expense={ props.expense }
      onSubmit={ ( oneExpense ) => {
        props.dispatch( editExpense( props.expense.id, oneExpense ) );
        props.history.push( '/' );
      }}
    />
    <button onClick={ () => {      
      const theId = props.expense.id;
      console.log( 'props expense', { id: theId } );
      props.dispatch( removeExpense( { id: theId } ) );
      props.history.push( '/' );
    }}>
      Remove
    </button>
  </div>
);

const mapStateToProps = ( state, props ) => {
  return {
    expense: state.expenses.find( ( oneExpense ) => oneExpense.id === props.match.params.id )
  };
};

export default connect( mapStateToProps )( EditExpensePage );
