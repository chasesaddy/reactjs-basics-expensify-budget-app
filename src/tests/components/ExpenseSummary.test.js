import React from 'react';
import { shallow } from 'enzyme';

import { ExpenseSummary } from '../../components/ExpenseSummary';

test( 'should show display for single expense', () => { 
  const shouldBe = 'Viewing 1 expense totalling $3.50';

  const wrapper = shallow(<ExpenseSummary expenseCount={ 1 } expensesTotal={ 350 } /> );
  const specificHtml = wrapper.find( 'div' ).at( 0 );

  expect( specificHtml.text() ).toBe( shouldBe );
  expect( wrapper ).toMatchSnapshot();
} );

test( 'should show display for multiple expenses', () => { 
  const shouldBe = 'Viewing 3 expenses totalling $81.40';

  const wrapper = shallow(<ExpenseSummary expenseCount={ 3 } expensesTotal={ 8140 } /> );
  const specificHtml = wrapper.find( 'div' ).at( 0 );

  expect( specificHtml.text() ).toBe( shouldBe );
  expect( wrapper ).toMatchSnapshot();
} );
