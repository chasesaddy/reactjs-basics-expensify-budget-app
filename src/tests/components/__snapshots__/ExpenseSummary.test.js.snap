// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should show display for multiple expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        3
      </span>
       
      expenses
       totalling 
      <span>
        $81.40
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should show display for single expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $3.50
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
